ethereum/spec/src/main/java/tech/pegasys/teku/spec/logic/versions/altair/statetransition/epoch/ValidatorStatusFactoryAltair.java

```diff
@@ -52,33 +53,86 @@ protected void processParticipation(
+            currentParticipationFlags, ParticipationFlags.TIMELY_SOURCE_FLAG_INDEX)) {
+          status.updateCurrentEpochSourceAttester(true);
+        }
+        if (miscHelpersAltair.hasFlag(
+            currentParticipationFlags, ParticipationFlags.TIMELY_TARGET_FLAG_INDEX)) {
+          status.updateCurrentEpochTargetAttester(true);
+        }
       }
     }
+  }
 
-    final SszList<SszByte> currentParticipation = state.getCurrentEpochParticipation();
-    for (int i = 0; i < currentParticipation.size(); i++) {
-      final SszByte participationFlags = currentParticipation.get(i);
-      if (miscHelpersAltair.hasFlag(
-          participationFlags.get(), ParticipationFlags.TIMELY_SOURCE_FLAG_INDEX)) {
-        statuses.get(i).updateCurrentEpochSourceAttester(true);
+  protected TotalBalances createTotalBalances(final List<ValidatorStatus> statuses) {
+    UInt64 currentEpochActiveValidators = UInt64.ZERO;
+    UInt64 previousEpochActiveValidators = UInt64.ZERO;
+    UInt64 currentEpochSourceAttesters = UInt64.ZERO;
+    UInt64 currentEpochTargetAttesters = UInt64.ZERO;
+    UInt64 previousEpochSourceAttesters = UInt64.ZERO;
+    UInt64 previousEpochTargetAttesters = UInt64.ZERO;
+    UInt64 previousEpochHeadAttesters = UInt64.ZERO;
+
+    for (ValidatorStatus status : statuses) {
+      final UInt64 balance = status.getCurrentEpochEffectiveBalance();
+      if (status.isActiveInCurrentEpoch()) {
+        currentEpochActiveValidators = currentEpochActiveValidators.plus(balance);
+      }
+      if (status.isActiveInPreviousEpoch()) {
+        previousEpochActiveValidators = previousEpochActiveValidators.plus(balance);
+      }
+
+      if (status.isSlashed()) {
+        continue;
+      }
+      if (status.isCurrentEpochSourceAttester()) {
+        currentEpochSourceAttesters = currentEpochSourceAttesters.plus(balance);
+      }
+      if (status.isCurrentEpochTargetAttester()) {
+        currentEpochTargetAttesters = currentEpochTargetAttesters.plus(balance);
+      }
+
+      if (status.isPreviousEpochSourceAttester()) {
+        previousEpochSourceAttesters = previousEpochSourceAttesters.plus(balance);
+      }
+      if (status.isPreviousEpochTargetAttester()) {
+        previousEpochTargetAttesters = previousEpochTargetAttesters.plus(balance);
       }
-      if (miscHelpersAltair.hasFlag(
-          participationFlags.get(), ParticipationFlags.TIMELY_TARGET_FLAG_INDEX)) {
-        statuses.get(i).updateCurrentEpochTargetAttester(true);
+      if (status.isPreviousEpochHeadAttester()) {
+        previousEpochHeadAttesters = previousEpochHeadAttesters.plus(balance);
       }
     }
+    return new TotalBalances(
+        specConfig,
+        currentEpochActiveValidators,
+        previousEpochActiveValidators,
+        currentEpochSourceAttesters,
+        currentEpochTargetAttesters,
+        previousEpochSourceAttesters,
+        previousEpochTargetAttesters,
+        previousEpochHeadAttesters);
   }
```

Comment:
This is the key change - almost identical to the phase0 version but performs each check independently instead of nesting checks under the `isCurrentEpochSourceAttester` or `isPreviousSourceAttester` checks.

Reply:
I think we can just make this change in the abstract base class so that we don't have to duplicate the logic here.

Reply:
Yeah took me a bit of reasoning but we only set the correct head flag if the target was correct so can skip the nesting here for both versions.  Good spot.